### ISSUE TITLE
Allow non-installation of vendored pep514tools on Unix

### ISSUE THREAD

**Issue by jayvdb** [sarugaku/pythonfinder#64]
It would be nice if the code and contracts with the user were such that Unix distros do not need to include the vendored pep514tools.

The main motivation is that it isnt released, so naming of the distro package is problematic, but it is also mostly unnecessary on Unix, and distros tend to like to de-vendor packages, so it would be a bit odd to create a separate package which is useless on that platform.  Better if the imports are guarded with `if os.name == "nt"` so it could be simply omitted on Unix distros.

But would that mean useful functionality is missing from the package when used on Unix ?

**Comment by techalchemy** [comment]
It is released.

**Comment by techalchemy** [comment]
Ah on pypi, sure. There isn’t really any way to avoid this until it is on pypi and the cost of inclusion is essentially 0; it is a 3 file library and is imported on Linux only. I do get the de-bundling concern and my answer is basically: don’t debundle it.

**Comment by jayvdb** [comment]
Ha, I am well aware of pypa perspective on de-bundling, and given the effort you have put into properly managing bundled packages it seems pointless to try to replicate the version management by the very small openSUSE Python maintainer team.  But I tried getting pipenv into openSUSE without debundling, but that didnt lead anywhere https://build.opensuse.org/request/show/674695 .

I saw that Fedora also partially de-bundled pipenv, so I didnt try to argue against it in openSUSE, as I have found the openSUSE maintainers rarely change their mind once making a suggestion.  I've almost finished fully debundling pipenv, but there are a few deps under pipenv which are bundling, such as pythonfinder, so I'm going to sort those out too for completeness and to try to make them happy.

**Comment by techalchemy** [comment]
Fair enough, I basically strongly advise against it but leave the issues open so I’m not really opposed, I just don’t have a good way to handle these specific things

Our vendoring scripts are messy over in pipenv at this point but they work somehow. Definitely due for a refactor. I basically maintain the ecosystem myself but the tooling allows our other maintainers to step in as needed.

**Comment by jayvdb** [comment]
Probably the easiest way to provide a guarantee that Unix doesnt need `pep514tools` is to have a travis job which deletes the vendored copy - the tests should all pass, and in fact they do currently.

Or `pep514tools` could be patched to raise an exception if it is used on Unix.  Or some other similar solution.

**Comment by techalchemy** [comment]
To put it bluntly, I'm probably going to keep vendoring the library. I don't build separate platform wheels for each release and the library in question isn't up on PyPI so the only way to guarantee its inclusion is to vendor it (yes I know that modern pep508 syntax allows flexibility but not everyone uses modern things).

I am essentially treating it as part of `pythonfinder` because of these factors, and while I understand the repackaging concerns, I am one person, and I really can't manage the redistribution process of each linux distro as well as the 15 libraries I am juggling

**Comment by jayvdb** [comment]
I agree with all of that.  And I typically help if I can, and issues are a way of understanding how I can to help.

In my latest post here, I suggested that the unix CI jobs (or a new job) removes the vendored copy of `pep514tools` (windows) as part of the job logic.  If the tests pass in those jobs, then the rest of the code isnt accidentally importing the vendored copy of `pep514tools` on Unix, which means it can safely be deleted by anyone who wants to (e.g. Unix OS vendors).

This is just a preventative measure to ensure that a Windows library doesnt accidentally get loaded when it isnt needed or useful.
I have already [done this for openSUSE](https://build.opensuse.org/package/view_file/devel:languages:python/python-pythonfinder/python-pythonfinder.spec?expand=1).  It works.  The local tests, and all the tests of related libraries which are needed by pipenv, and pipenv, all pass.  If there is any codepath where `pep514tools` is imported on Unix, it must not be very important, as nothing is using that codepath afaics.

So all that should be required is adding `rm -r .../pep514tools` in a Unix job.  Unless I hear violet objections, I'll propose that in a PR.

**Comment by uranusjr** [comment]
One thing I don’t quite follow is *why* this is necessary. AFAICT the import is always guarded, what is wrong with just packaging the library as-is without de-vendoring? Is there a rule in OpenSUSE (or other distros) that forbids this?

**Comment by jayvdb** [comment]
It might be guarded now, but is it going to always be the case?
That is why I opened this issue, and mentioned 'contracts with the user' (of this library).  Ideally there is protection here from the vendored package ever being accidentally imported on Unix.  e.g. in five years from now, after the repo has been moved to an org, and active maintainers switched a few times.

Distros try to prevent each package being a special case, because there will be so many packages all claiming a reason why they should be exempt from the de-bundling/de-vendoring rules, and the process of creating the distro requires too many lawyers to check all the special cases, and engineering manpower to control the complexity of the constant stream security vulnerabilities, support queries, etc.

Distros do have rather strong language in their guidelines about debundling, especially bundling of system libraries, but the same principles apply throughout, and are applied throughout with varying intensity.

I am sure that the case for not de-vendoring pep514tools from pythonfinder would be easy to write persuasively.  The risks of a separate but dormant python codebase inside a python package are much lower than that of a hidden C library.  But why bother writing a justification; it would be repeatedly revisited throughout the life of the package; is it worth it?

pythonfinder is a really simple case where de-vendoring is easy, a tiny engineering effort up-front, so it would be silly for this to be a special case of a single package with two codebases inside it, needing extra metadata to describe that, especially the licenses (both MIT, but different authors).  My openSUSE package deletes the pep514tools so that my package will break with tests failing if a future release tries to use pep514tools.  That should make it easy to maintain - i.e. update each time there is a new release, which means I can spend more time on the other packages which are more troublesome, and this one can get upgraded on almost autopilot.  I only need to check each release does not include strange new unexpected files.  A CI enforcement at the origin (i.e. here) should make it impossible for the problem to occur, as the problem will get caught here, before hitting PyPI and thus pushed into the upgrade queues of distros.

OTOH, personally I believe pip and pipenv should be special cases, where the engineering effort to de-bundle and QA each release is much higher than the effort to check the bundling in each release is sane and there are no surprises that could cause legal problems (e.g. the GPL cursor library which had to be replaced).  pip is allowed to include bundled packages in openSUSE, but it isnt in Fedora or Debian last time I checked.  I tried that argument wrt pipenv, and openSUSE maintainers dont appear to agree.
